designs/0006-labels.md

```diff
@@ -0,0 +1,289 @@
+* kubernetes.azure.com/ebpf-dataplane
+* kubernetes.azure.com/cluster-health-monitor-checker-synthetic (due to its high usage...) **TODO:** Learn what this is
+
+Based on usage data, the following `kubernetes.azure.com` labels have some usage today in `NodePool` labels or requirements.
+
+Most usage
+
+* kubernetes.azure.com/ebpf-dataplane
+* kubernetes.azure.com/mode
+* kubernetes.azure.com/cluster-health-monitor-checker-synthetic
+* kubernetes.azure.com/scalesetpriority
+
+Less usage
+* kubernetes.azure.com/agentpool
+* kubernetes.azure.com/accelerator
+* kubernetes.azure.com/agentpool-family
+* kubernetes.azure.com/os-sku
+* kubernetes.azure.com/aksnodeclass
+* kubernetes.azure.com/storageprofile
+* kubernetes.azure.com/storagetier
+
+We already block usage of the `karpenter.azure.com` and `karpenter.sh` namespaces except for certain labels we understand. This will not change.
+
+We will also block all `kubernetes.azure.com` namespaced labels except for those called out above as either well known labels, or allowed.
+
+### Decision 5: How will we get labels that Karpenter doesn't currently write set on nodes?
+
+#### Option A: Wait for Machine API migration
+
+Migrating to the machine API will automatically get all missing labels set on VMs allocated via the machine API.
+
+#### Option B: Add some/all labels to our current label writing mechanism
+
+There are some labels that Karpenter sets but that aren't schedulable, defined at [AddAgentBakerGeneratedLabels](https://github.com/Azure/karpenter-provider-azure/blob/main/pkg/providers/imagefamily/labels/labels.go#L23).
+This has already been accounted for in the tables above. We could add more labels there.
+
+#### Conclusion: Wait for machine API for complete parity
+
+Machine API is not that far away, and we already write most critical labels. Some additional labels such as `kubernetes.azure.com/os-sku` and `kubernetes.azure.com/fips_enabled` will be enabled for scheduling
+(and thus be written to the nodes) based on decision 4, but other labels and full label parity will come with the machine API migration.
+
+## FAQ
+
+### How do the labels get onto the Node?
+
+* Karpenter core writes all NodeClaim labels to the node, see: [syncNode](https://github.com/kubernetes-sigs/karpenter/blob/38e728c99b530f660f685d53d01f2e9ec9696668/pkg/controllers/nodeclaim/lifecycle/registration.go#L128).
+* Karpenter core translates most requirements from the NodeClaim spec into labels, see [PopulateNodeClaimDetails](https://github.com/kubernetes-sigs/karpenter/blob/38e728c99b530f660f685d53d01f2e9ec9696668/pkg/controllers/nodeclaim/lifecycle/launch.go#L129)
+    * Note that `.Labels()` does exclude multi-valued requirements and labels defined in `WellKnownLabels` (registered by the cloud provider). See [Labels()](https://github.com/kubernetes-sigs/karpenter/blob/38e728c99b530f660f685d53d01f2e9ec9696668/pkg/scheduling/requirements.go#L270).
```

Comment:
Exclusion: `.Labels()` excludes "restricted" labels, which include `WellKnownLabels`, but also include those in `RestrictedLabelDomains` ("prohibited by kubelet or reserved by karpenter", e.g. "kubernetes.io", or "karpenter.sh".) - unless under `LabelDomainExceptions` (e.g. "not used in a context where they may be passed as argument to kubelet", e.g. "node.kubernetes.io").

Reply:
You're right this isn't actually accurate.

Have updated to:
```
    * Note that `.Labels()` excludes labels defined in `WellKnownLabels`, including both those registered by the cloud provider and core. It also excludes labels defined in `RestrictedLabelDomains`, which are
      labels that are prohibited by Kubelet or reserved by Karpenter, except for certain sub-domains within the restricted domains defined by `LabelDomainExceptions`.
```